### PR TITLE
dependency: avoid expanding same dep twice

### PR DESCRIPTION
**Pull request by vladshablinsky**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew tests` with your changes locally?

## description

Consider the case when with these formulae:
`a`, `b`, `c`, `d`, `e`.

* `a` dependencies: `b`
* `b` dependencies: `c, e`
* `c` dependencies: `d`
* `e` dependencies: `c`

If we're expanding dependencies using post-order traversal

* expand left subtree
* expand right subtree
* display current node

we're doing some unnecessary computations for `c` and `d` which are
displayed in the parantheses:

```
         a
        /
       b
      / \
     c   e
    /     \
   d      (c)
          /
        (d)
```

Let's expand a's dependencies:

1. `a`, call recursively for `b`
2. `b`, call recursively for `c`
3. `c`, call recursively for `d`
4. `d`, push `d` to dependencies list, which now gets `[d]`, go up to
`c`
5. `c`, push `c` to dependencies list, which now gets `[d, c]`, go up to `b`
6. `b`, call recursively for `e`
7. _`e`, call recursively for `c` <-- unnecessary call_
8. _`c`, call recursively for `d` <-- unnecessary call_
9. `d`, dependencies list already has `[d, c]`, go up to `c`
10. `c`, go up to `e`, push `e` to dependencies list, which now gets `[d, c,
e]`, go up to `b`
11. `b`, push `b` to dependencies list, which now gets `[d, c, e, b]`, go up
to a
12. `a`, push `a` to dependencies list, which now gets `[d, c, e, b, a]`

## tl;dr

Let's skip expanding subtrees of dependencies we have already expanded
once, since they are the same as we have already expanded once.

### CODE REVIEW

**Comment by MikeMcQuaid** [comment]
> Let's skip expanding subtrees of dependencies we have already expanded
> once, since they are the same as we have already expanded once.

Just to check the obvious thing that popped into my mind: how does this handle e.g. dependencies with options? Similarly, do you have any benchmarks that show a noticeable performance increase? I tend to favour only caching for measurably performance impact.

**Comment by vladshablinsky** [comment]
> Just to check the obvious thing that popped into my mind: how does this handle e.g. dependencies with options?

Well, the results of `expand(dep.to_formula, &block)`can differ if we pass same `&block` and same `dep` but with different options. Will need to think about it a bit more.

> I tend to favour only caching for measurably performance impact.

Good point. I think if we cache it properly, we'll able to reuse it if we pass a list of formulae to install to `brew`.

> Similarly, do you have any benchmarks that show a noticeable performance increase? 

It's not that difficult to create an artificial example where it should be e.g. twice more efficient, however I don't have benchmarks.

**Comment by MikeMcQuaid** [comment]
> It's not that difficult to create an artificial example where it should be e.g. twice more efficient, however I don't have benchmarks on some real data.

I'd like to gently suggest that caching will introduce edge cases and without user-visible performance increases I'm probably 👎 on it. Performance tuning/caching is valuable, though, but I'd do so using `ruby-prof --printer=call_stack` or [`brew prof`](https://github.com/Homebrew/homebrew-dev-tools/blob/master/cmd/brew-prof.rb) on a particular command we know has low performance i.e. `brew uses --recursive`.

**Comment by MikeMcQuaid** [comment]
Passing on this for now, sorry @vladshablinsky 😭

**Comment by vladshablinsky** [comment]
@MikeMcQuaid no worries. It really requires some testing performed and probably rethinking.
